File: uncertaintext.py
#!/usr/bin/env python
"""Uncertaintext python prototype

Exploring the basic idea of representing uncertain values with samples from the
underlying distribution
"""

import click
from random import gauss
import time


@click.command()
@click.option('--mu', type=float, default=0., help='Mean, parameter to gaussian distribution')
@click.option('--sigma', type=float, default=1., help='Standard deviation, parameter to gaussian distribution')
@click.option('--fps', type=int, default=10, help='Target update rate, in "frames per second"')
@click.option('--duration', type=float, default=5, help='Demo duration in seconds')
def cli(mu: float, sigma: float, fps: int, duration: float):
    """Display uncertaintext representation of the specified distribution
    """
    delay = 1. / fps
    exit_time  = time.monotonic() + duration
    base = f'{mu:.2f} \u00B1 {2*sigma:.2f}'

    while (now := time.monotonic()) < exit_time:
        click.echo(f'{base} ({gauss(mu, sigma): .2f})\r', nl=False)
        time.sleep(max(0, (now + delay) - time.monotonic()))
    click.echo()
        

if __name__ == '__main__':
    cli()

File: uncertaintext.js
// Uncertaintext: uncertain text for uncertain numbers
//
// Uncertaintext provides dynamic styling for representing probability distributions. The core idea is
// to express probabalistic values as series of random samples from the distribution, rather than the
// usual summary statistics. The aim is to give readers a "gut feel" for the uncertainty, and have
// a little fun doing it.
//
// To use, define your distributions as <span> elements with the following attributes:
//  * class=uncertaintext: required, marks the span element for uncertaintext to find and update
//  * data-uct-distrib=[string], required, name of the distribution to sample from, currently supported
//      names are: normal.
//  * data-uct-mu=[float]: required for normal distribution, mean
//  * data-uct-sigma=[float]: required for normal distribution, standard deviation
//  * data-uct-format=[string]: optional, printf-style format string to apply to the sample,
//      see https://github.com/d3/d3-format
//  * data-uct-fps=[int]: optional, update frequency in "frames" per second
//
// dependencies:
//  * d3-format: https://github.com/d3/d3-format
//  * d3-random: https://github.com/d3/d3-random
////


// retrieve dataset attribute or fail if not set
function required_data(element, name) {
    return element.dataset[name]
}


// retrieve dataset attribute or default if not set
function optional_data(element, name, fallback) {
    if (name in element.dataset) {
        return element.dataset[name]
    }
    return fallback
}


// update element with formatted sample from the distribution
function update_sample(target, sampler, sample_format) {
    target.innerHTML = sample_format(sampler());
}


// initialize all uncertaintext elements on the page
function uncertaintext() {

    let targets = document.getElementsByClassName("uncertaintext")

    for (let i = 0; i < targets.length; i++) {

        let target = targets[i];

        try {

            let sampler = null; 

            // distribution definition (required)
            let distribution_name = required_data(target, 'uctDistrib');
            
            // sampling function (required)
            if (distribution_name === 'normal') {
                let mu    = parseFloat(required_data(target, 'uctMu'));
                let sigma = parseFloat(required_data(target, 'uctSigma'));
                sampler = d3.randomNormal(mu, sigma);
            
            // TODO: add another distribution or two
            
            } else {
                console.log('No support for distribution: "%s"', distribution_name);
            }

            // format specifications (optional) 
            let sample_format = d3.format(optional_data(target, 'uctFormat', ' .2f'));

            // update interval (optional) 
            let fps = optional_data(target, 'uctFps', 5);
            let delay_ms = 1. / fps * 1000.

            // define update function
            let updater = function() {
                target.innerHTML = sample_format(sampler());
            };

            // call once (to setup page an trigger errors where we catch them), then set to repeat
            updater();
            setInterval(updater, delay_ms);        

        } catch(err) {
            // indicate error, and continue on to the next element 
            console.log(err);
            target.innerHTML = '[error]';
        }    
    } 

}
